### PR TITLE
PTQ support for OVModelForFeatureExtraction and OVSentenceTransformer

### PR DESCRIPTION
**Pull request by nikita-savelyevv**
# What does this PR do?

Add support for full quantization of feature-extraction models which are implemented through `OVModelForFeatureExtraction`  and `OVSentenceTransformer` depending on the library.

I've compared embeddings computed for the text below between FP32 and INT8 compressed/quantized models. Similarity between the embeddings is the cosine similarity. Results are below.

Text: "OpenVINO is an open-source software toolkit for optimizing and deploying deep learning models.  It enables programmers to develop scalable and efficient AI solutions with relatively few lines of code. It supports several popular model formats[2] and categories, such as large language models, computer vision, and generative AI."

| Model id                                                       | Model Class                        | Similarity INT8 WC | Similarity INT8 PTQ |
|-----------------------------------------------------------|-----------------------------|--------|---------|
| BAAI/bge-large-en-v1.5                                    | OVModelForFeatureExtraction | 0.9997 | 0.9964  |
| BAAI/bge-m3                                               | OVModelForFeatureExtraction | 0.9939 | 0.9938  |
| BAAI/bge-small-en-v1.5                                    | OVModelForFeatureExtraction | 0.9998 | 0.9979  |
| cross-encoder/ms-marco-MiniLM-L-6-v2                      | OVModelForFeatureExtraction | 0.9999 | 0.9995  |
| cross-encoder/ms-marco-MiniLM-L-6-v2                      | OVModelForFeatureExtraction | 0.9999 | 0.9995  |
| sentence-transformers/all-MiniLM-L12-v2                   | OVSentenceTransformer       | 0.9996 | 0.9967  |
| sentence-transformers/all-MiniLM-L6-v2                    | OVSentenceTransformer       | 0.9995 | 0.9971  |
| sentence-transformers/all-distilroberta-v1                | OVSentenceTransformer       | 0.9992 | 0.9946  |
| sentence-transformers/all-mpnet-base-v2                   | OVSentenceTransformer       | 0.9996 | 0.9870  |
| sentence-transformers/all-mpnet-base-v2                   | OVSentenceTransformer       | 0.9996 | 0.9870  |
| sentence-transformers/multi-qa-MiniLM-L6-cos-v1           | OVSentenceTransformer       | 0.9996 | 0.9979  |
| sentence-transformers/paraphrase-MiniLM-L6-v2             | OVSentenceTransformer       | 0.9994 | 0.9951  |
| sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2 | OVSentenceTransformer     | 0.9998 | 0.9987  |
| sentence-transformers/paraphrase-multilingual-mpnet-base-v2 | OVSentenceTransformer     | 0.9998 | 0.9987  |

PTQ was applied using "wikitext2" dataset and 128 number of samples.

Ticket 165770

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/intel/openvino/quantization.py [hunk 1]
```diff
@@ -744,6 +784,53 @@ def disable_progress_bar(self, disable: bool = True) -> None:
...
+
+        self.model.compile()
+
+        num_samples = quantization_config.num_samples or 128
+        calibration_data = []
+        try:
+            self.model.request = InferRequestWrapper(
+                self.model.request,
+                calibration_data,
+                inference_result_mock={
```
Comment (eaidova) [inline optimum/intel/openvino/quantization.py:803]: why this empty np arrays are needed?

File: optimum/intel/openvino/quantization.py [hunk 2]
```diff
@@ -744,6 +784,53 @@ def disable_progress_bar(self, disable: bool = True) -> None:
...
+                    "token_embeddings": np.empty((1,), np.float32),
+                    "sentence_embedding": np.empty((1,), np.float32),
+                },
+            )
+
+            tokenizer = None
+            pbar = tqdm(total=num_samples, desc="Collecting calibration data")
+            for item in dataset:
+                sentence = item["text"].strip()[:seq_len]
+                if len(sentence) < seq_len:
```
Comment (eaidova) [inline optimum/intel/openvino/quantization.py:814]: what about using padding and truncation during tokenization instead of filtering small strings by number of words. Besides that number of symbols in string is not always equal length of tokenized input (some letter may be represented by 2-4 tokens, some subwords may be represented by 1 token), this approach does not looks robust

**Comment by MaximProshin** [comment]
@nikita-savelyevv , it seems the following models are also covered from WCR scope: 
https://huggingface.co/cross-encoder/ms-marco-MiniLM-L-6-v2
https://huggingface.co/sentence-transformers/all-distilroberta-v1
or not?

File: optimum/intel/openvino/quantization.py (hunk 1 above)
Comment (nikita-savelyevv) [reply optimum/intel/openvino/quantization.py:803]: These are for the forward method to be able to post-process the model outputs, for example  https://github.com/huggingface/optimum-intel/blob/main/optimum/intel/openvino/modeling_sentence_transformers.py#L61

**Comment by nikita-savelyevv** [comment]
> @nikita-savelyevv , it seems the following models are also covered from WCR scope: https://huggingface.co/cross-encoder/ms-marco-MiniLM-L-6-v2 https://huggingface.co/sentence-transformers/all-distilroberta-v1 or not?

Thanks, added

File: optimum/intel/openvino/quantization.py (hunk 2 above)
Comment (nikita-savelyevv) [reply optimum/intel/openvino/quantization.py]: Thanks for the suggestion! I've added truncation based on number of tokens. I don't think padding is a good idea, because some inputs will contain too few actual non-padded tokens.

File: optimum/intel/openvino/quantization.py [hunk 3]
```diff
@@ -744,6 +784,53 @@ def disable_progress_bar(self, disable: bool = True) -> None:
...
+                    continue
+
+                if isinstance(self.model, OVModelForFeatureExtraction):
+                    tokenizer = tokenizer or AutoTokenizer.from_pretrained(quantization_config.tokenizer)
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply optimum/intel/openvino/quantization.py]: Added

File: optimum/commands/export/openvino.py [hunk 4]
```diff
@@ -448,10 +452,18 @@ def run(self):
...
                 model_cls = OVModelForVisualCausalLM
-            else:
+            elif "automatic-speech-recognition" in task:
                 from optimum.intel import OVModelForSpeechSeq2Seq
 
                 model_cls = OVModelForSpeechSeq2Seq
+            elif task.startswith("feature-extraction") and library_name in ["transformers", "sentence_transformers"]:
+                from ...intel import OVModelForFeatureExtraction, OVSentenceTransformer
+
+                model_cls = OVModelForFeatureExtraction if library_name == "transformers" else OVSentenceTransformer
```
Comment (IlyasMoutawwakil) [inline optimum/commands/export/openvino.py:462]: why importing both here ? attempts importing `OVSentenceTransformer` even if `sentence-transformers` is not the lib

Comment (nikita-savelyevv) [reply optimum/commands/export/openvino.py:462]: Do you suggest something like this?

```python
elif task.startswith("feature-extraction") and library_name == "transformers":
    from ...intel import OVModelForFeatureExtraction

    model_cls = OVModelForFeatureExtraction
elif task.startswith("feature-extraction") and library_name == "sentence_transformers":
    from ...intel import OVSentenceTransformer

    model_cls = OVSentenceTransformer
```

File: tests/openvino/test_quantization.py [hunk 5]
```diff
@@ -451,30 +482,30 @@ def test_ov_model_static_quantization_with_auto_dataset(
...
-            elif model_cls == OVModelForCausalLM:
-                submodels = [ov_model]
-
+            elif model_cls in (OVModelForCausalLM, OVModelForFeatureExtraction):
                 tokenizer = AutoTokenizer.from_pretrained(model_id)
                 if tokenizer.pad_token is None:
                     tokenizer.pad_token = tokenizer.eos_token
                 tokens = tokenizer("This is a sample input", return_tensors="pt")
-                outputs = ov_model(**tokens)
-                self.assertTrue("logits" in outputs)
```
Comment (IlyasMoutawwakil) [inline tests/openvino/test_quantization.py:472]: why is this removed ?

File: tests/openvino/test_quantization.py [hunk 6]
```diff
@@ -451,30 +482,30 @@ def test_ov_model_static_quantization_with_auto_dataset(
...
-                for x in (OVStableDiffusionPipeline, OVStableDiffusionXLPipeline, OVLatentConsistencyModelPipeline)
+                ov_model(**tokens)
+            elif model_cls in (
+                OVStableDiffusionPipeline,
+                OVStableDiffusionXLPipeline,
+                OVLatentConsistencyModelPipeline,
             ):
-                submodels = ov_model.ov_submodels.values()
+                # TODO: add inference check
+                pass
```
Comment (IlyasMoutawwakil) [inline tests/openvino/test_quantization.py:506]: is this todo for another PR ?

File: optimum/commands/export/openvino.py (hunk 4 above)
Comment (IlyasMoutawwakil) [reply optimum/commands/export/openvino.py:462]: yes

File: tests/openvino/test_quantization.py (hunk 6 above)
Comment (nikita-savelyevv) [reply tests/openvino/test_quantization.py:506]: Yes

File: optimum/intel/openvino/modeling_sentence_transformers.py [hunk 7]
```diff
@@ -106,6 +110,15 @@ def _from_pretrained(
             **kwargs,
         )
 
+        if quantization_config is not None:
```
[the remark this answers is not in the document]
Comment (nikita-savelyevv) [reply optimum/intel/openvino/modeling_sentence_transformers.py:113]: Thanks for noticing, fixed

File: tests/openvino/test_quantization.py (hunk 5 above)
Comment (IlyasMoutawwakil) [reply tests/openvino/test_quantization.py:472]: at least it runs the quantized model.. isn't the idea behind this test is to load with quantization and run inference sanity checks ?

File: tests/openvino/test_quantization.py (hunk 6 above)
Comment (IlyasMoutawwakil) [reply tests/openvino/test_quantization.py:506]: why not now ? it's as simple as `ov_model(prompt="a text to image prompt")`

File: tests/openvino/test_quantization.py (hunk 5 above)
Comment (IlyasMoutawwakil) [reply tests/openvino/test_quantization.py:472]: okay I see the `ov_model(**tokens)` afterwards

Comment (nikita-savelyevv) [reply tests/openvino/test_quantization.py:472]: The model is still run, the only thing missing is `self.assertTrue("logits" in outputs)`

File: tests/openvino/test_quantization.py (hunk 6 above)
Comment (nikita-savelyevv) [reply tests/openvino/test_quantization.py]: Indeed, added :slightly_smiling_face:

File: optimum/commands/export/openvino.py (hunk 4 above)
Comment (nikita-savelyevv) [reply optimum/commands/export/openvino.py]: Changed

File: optimum/intel/openvino/utils.py [hunk 8]
```diff
@@ -137,8 +137,12 @@
     "automatic-speech-recognition-with-past": "OVModelForSpeechSeq2Seq",
 }
 
+CAUSAL_LANGUAGE_DATASETS = ["wikitext2", "c4", "c4-new", "auto"]
 
-LANGUAGE_DATASETS = ["wikitext2", "c4", "c4-new", "auto"]
+LANGUAGE_DATASETS = {
```
Comment (echarlaix) [inline optimum/intel/openvino/utils.py:142]: could make sense to add the "PREDEFINED" prefix to match  the rest
```suggestion
PREDEFINED_LANGUAGE_DATASETS = {
```

File: optimum/intel/openvino/configuration.py [hunk 9]
```diff
@@ -816,11 +822,20 @@ def post_init(self):
...
             stable_diffusion_datasets = list(PREDEFINED_SD_DATASETS.keys())
-            if self.dataset not in LANGUAGE_DATASETS + speech_to_text_datasets + stable_diffusion_datasets:
+            language_datasets = list(LANGUAGE_DATASETS.keys())
+            if (
+                self.dataset
+                not in language_datasets
+                + CAUSAL_LANGUAGE_DATASETS
+                + speech_to_text_datasets
+                + stable_diffusion_datasets
+            ):
```
Comment (echarlaix) [inline optimum/intel/openvino/configuration.py:832]: not very important but the resulting list will have some redundant values

**Review by echarlaix — APPROVED** [review]
LGTM!

File: optimum/intel/openvino/utils.py (hunk 8 above)
Comment (nikita-savelyevv) [reply optimum/intel/openvino/utils.py]: Thanks! Renamed

File: optimum/intel/openvino/configuration.py (hunk 9 above)
Comment (nikita-savelyevv) [reply optimum/intel/openvino/configuration.py]: Converted to sets.
